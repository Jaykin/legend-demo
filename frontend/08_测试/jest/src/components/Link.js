/**
 * Created by Administrator on 2017/1/19.
 */
import React from 'react';

const STATUS = {
    NORMAL: 'normal',
    HOVERED: 'hovered',
};

class Link extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            className:STATUS.NORMAL
        }

        this._onMouseEnter = this._onMouseEnter.bind(this);
        this._onMouseLeave = this._onMouseLeave.bind(this);
    }

    _onMouseEnter() {
        this.setState({className: STATUS.HOVERED});
    }

    _onMouseLeave() {
        this.setState({className: STATUS.NORMAL});
    }

    render() {
        return (
            <a
                className={this.state.className}
                href={this.props.page || '#'}
                onMouseEnter={this._onMouseEnter}
                onMouseLeave={this._onMouseLeave}>
                百度一下会死啊！
            </a>
        );
    }
}

export default Link;