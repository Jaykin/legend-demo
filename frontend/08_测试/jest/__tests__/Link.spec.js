/**
 * Created by Administrator on 2017/1/20.
 */
import React from 'react';
import Link from '../src/components/Link';
import renderer from 'react-test-renderer';


it('正确渲染', () => {
    const tree = renderer.create(
        <Link page="https://www.baidu.com">百度</Link>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('链接不设置地址', () => {
    const tree = renderer.create(
        <Link>百度</Link>
    ).toJSON();
    expect(tree).toMatchSnapshot();
})

it('properly escapes quotes', () => {
    const tree = renderer.create(
        <Link>{'"Facebook" \\\'is \\ \'awesome\''}</Link>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

// Snapshot Testing
test('Link changes the class when hovered', () => {
    const component = renderer.create(
        <Link page="http://www.facebook.com">Facebook</Link>
    );
    let tree = component.toJSON();
    //console.log(component);
    //console.log(tree);
    expect(tree).toMatchSnapshot();     // 截获虚拟DOM？？
    /*tree结构
    * {
    *    type: 'a',
         props: {
             className: 'normal',
             href: 'http://www.facebook.com',
             onMouseEnter: [Function: bound _onMouseEnter],
             onMouseLeave: [Function: bound _onMouseLeave]
         },
         children: [ '百度一下会死啊！' ] }
    * */


    // 手动触发回调
    tree.props.onMouseEnter();
    // re-rendering(重新绘制)
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // 手动触发回调
    tree.props.onMouseLeave();
    // re-rendering(重新绘制)
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});