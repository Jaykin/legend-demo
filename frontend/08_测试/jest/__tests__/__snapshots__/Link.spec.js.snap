exports[`test Link changes the class when hovered 1`] = `
<a
  className="normal"
  href="http://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  百度一下会死啊！
</a>
`;

exports[`test Link changes the class when hovered 2`] = `
<a
  className="hovered"
  href="http://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  百度一下会死啊！
</a>
`;

exports[`test Link changes the class when hovered 3`] = `
<a
  className="normal"
  href="http://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  百度一下会死啊！
</a>
`;

exports[`test properly escapes quotes 1`] = `
<a
  className="normal"
  href="#"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  百度一下会死啊！
</a>
`;

exports[`test 正确渲染 1`] = `
<a
  className="normal"
  href="https://www.baidu.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  百度一下会死啊！
</a>
`;

exports[`test 链接不设置地址 1`] = `
<a
  className="normal"
  href="#"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  百度一下会死啊！
</a>
`;
